plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/ConstraintContentProposalProvider.java

```diff
@@ -94,108 +92,134 @@ public ConstraintContentProposalProvider(Collection<String> featureNames) {
-			words[LAST] = "(";
-		}
-		if (words[LAST].endsWith(")")) {
-			words[LAST] = ")";
-			if (contents.charAt(posMarker) == ')') {
-				words[LAST] = ") ";
-			}
+			// the current typed word
+			final String currentWord = contents.substring(posMarker);
 
-		}
-		if (words[CURRENT].endsWith(")")) {
-			words[LAST] = ")";
-			words[CURRENT] = "";
+			if (quotationMark) {
+				posMarker--;
+			}
+			contents = contents.substring(0, posMarker);
+			contents = contents.trim();
 
+			// text before current word
+			if (contents.endsWith("\"")) {
+				return new ProposalContext(true, currentWord, quotationMark);
+			} else {
+				final String lastWord = contents.substring(contents.lastIndexOf(' ') + 1);
+				return new ProposalContext(!Operator.isOperatorName(lastWord) && !lastWord.isEmpty(), currentWord, quotationMark);
+			}
 		}
-
-		return words;
 	}
 
 	/**
-	 *
-	 * @param wordBefore
-	 *
-	 * @param features set of features
-	 * @return List of proposals, either operators or feature names
+	 * @param context context of proposal
+	 * @param features set of all features
+	 * @return List of proposals, either operators or feature names. Contains only proposals consistent with the previous text, but may contain proposals
+	 *         inconsistent with current word.
 	 */
-	private static List<ContentProposal> getProposalList(String wordBefore, Collection<String> features) {
+	private static List<ContentProposal> getProposalList(ProposalContext context, Collection<String> features) {
 
 		final ArrayList<ContentProposal> proposals = new ArrayList<ContentProposal>();
 		final ArrayList<String> featureList = new ArrayList<String>(features);
 		Collections.sort(featureList, String.CASE_INSENSITIVE_ORDER);
 
 		final Collection<String> operatorNamesInFeatures = Features.extractOperatorNamesFromFeatuers(features);
 
-		// TODO: Add binary operators only iff their appearance makes sense in content proposal
-		// Example:
-		// Show "and" for "A |"
-		// Hide "and" for "A and |"
-		proposals.add(new ContentProposal("and"));
-		proposals.add(new ContentProposal(IFF));
-		proposals.add(new ContentProposal(IMPLIES));
-		proposals.add(new ContentProposal("or"));
-
-		// TODO: Add binary operators only iff their appearance makes sense in content proposal
-		// Example:
-		// Show NOT for "A implies |"
-		// Hide NOT for "A |"
-		proposals.add(new ContentProposal(NOT));
-
-		// TODO: Add features only iff a feature name is valid in context
-		// Example:
-		// Show feature for "A implies |"
-		// Hide features for "A |"
-		for (final String s : featureList) {
-			proposals.add(new ContentProposal(s + (operatorNamesInFeatures.contains(s.trim()) ? " " + Features.FEATURE_SUFFIX : "")));
+		if (context.featureBefore) {
+			if (!context.quotationMark) {
+				// Add binary operators only iff their appearance makes sense in content proposal
+				// Example:
+				// Show "and" for "A |"
+				// Hide "and" for "A and |"
+				proposals.add(new ContentProposal("and"));
```

Comment:
Inconsistent usage of constants. Please also use constant strings for "and" and "or"